2186 csv exports are off by one. users are saying they have quantity but not really

dropkickfish opened the pull request:
Resolves #2186 

### Description
CSV Exports for distributions resulted in a CSV which had mismatched columns, due in part to additional items with zero quantity being added to the distribution.

This PR modifies the ExportPartnerDistributionsService object to also serve distributions as well as partners, and to take optional headers which will allow for more customisable CSV exports in the future. I've included the optional headers in the controller for this reason, but they could be limited to each model instead if we don't want to introduce optional headers for the CSV export and want to define those differently for each model.

I chose to modify this existing service object instead of creating a new one since 
1. I felt that it might be a bit closer to DRY principles
2. Following a [Slack discussion](https://rubyforgood.slack.com/archives/C6WLZL0DD/p1618483827085600) with Stanley, this better fits the purpose of a service object as opposed to having two different ones.

This is a WIP since I have not yet written a test, but early manual testing yields positive results with the correct CSV output.

### Type of change

* Bug fix (non-breaking change which fixes an issue)

### How Has This Been Tested?
It hasn't yet been tested beyond manual testing. Further testing is required.

dropkickfish commented:
@edwinthinks Would you mind re-running the tests, please? I've added a basic spec and everything's green on my machine, although I'd be open to ideas on how to improve that (I should also be on the call this evening)

edwinthinks commented:
Hey @dropkickfish thanks for working on this PR. I have some main suggestions/requests that we can go over today in the coding session. 
- I don't think we need to include `optional_headers` since there is no requirement for this yet. My stance on this is don't add the customization until we need it. We can always add it later if we want it.
- The logic for setting the header and aligning the item content to the seems to be missing or can be improved for more clarity. My approach would have been to build a method that constructs the headers based on the distribution items in alphabetical order and then in each row finds each corresponding column /w the Item.
- Adding a failing spec first to ensure that the code we do change will fix it. (TDD approach here)

I hope we can sit down together and work on this today.

dropkickfish commented:
Following feedback, I've 
- reverted the ExportPartnerDistributionsService and removed any references to additional headers
- created a new service object - ExportDistributionsCSVService
- updated specs accordingly for the above
- utilized [Set](https://ruby-doc.org/stdlib-2.7.1/libdoc/set/rdoc/Set.html) to check for duplicate headers instead of using array methods
- added logic to set nil values on items to zero in order to make csv exports compatible with more services (e.g. SQL)
- added logic to set nil values on Agency Representative (the only column value expecting a string which can populate the csv with a nil value) to "None" by default for the same reason

I was having some trouble running specs locally so I've been a bit cheeky and pushed this branch to run here.

Where I think this can be improved:
- I feel like the specs could probably be more robust
- I would like to be able to refactor the logic for the nil values more readable

edwinthinks commented:
@dropkickfish thanks for working on this! I should be able to test and review this out today.

edwinthinks commented on app/controllers/distributions_controller.rb:
```diff
@@ -57,7 +57,9 @@ def index
       format.html
-      format.csv { send_data Distribution.generate_csv(@distributions, @items.collect(&:name).sort), filename: "Distributions-#{Time.zone.today}.csv" }
+      format.csv do
+        send_data Exports::ExportDistributionsCSVService.new(distribution_ids: @distributions.map(&:id)).generate_csv, filename: "Distributions-#{Time.zone.today}.csv"
```
Removed the class method `generate_csv` since I think it complicates and makes it unclear how or what service object it is using.

edwinthinks commented on app/services/exports/export_distributions_csv_service.rb:
```diff
@@ -0,0 +1,85 @@
+    end
+
+    def build_row_data(distribution)
+      # Maybe utilize hash instead of array. (scott add comments that make senses)
```
@scooter-dangle hey there I think its ready to be modified based on your suggestion to use hash and rely on the headers to keep the order.

scooter-dangle replied:
I made that change and also optimized the index lookup.

scooter-dangle commented:
Note: I also rebased this onto `main`.
